### PR TITLE
use client/server tracers from com.github.kristofa.brave to annotate …

### PR DESCRIPTION
**Pull request by edwardsoo-ss**
…and flush kafka spans

Today if the kafka producer span is not sampled, the kafka consumer span will still sometimes annotate a single "sr", resulting in some weird looking spans with NaN duration on the UI (e.g. https://tracingd.smartthingstlm.com/traces/4114d36c7a484c87)

this PR change will make the consumer report "sr" only if the consumer record contains a trace context that is sampled. also switch to use tracers from `com.github.kristofa.brave` to flush spans

@Bijnagte @llinder please take a look

### CODE REVIEW

**Comment by llinder** [comment]
@edwardsoo-ss can you elaborate on the change from Brave v4 api to v3 api?  Did you run into an issue with the Brave v4 API?  If so I would like to get more information so we can escalate or provide feedback on the v4 API.

**Comment by Bijnagte** [comment]
👀

**Comment by edwardsoo-ss** [comment]
I actually prefer v4, namely that it fakes unsampled span with `brave.NoopSpan` so I never had to worry about NPE from created spans.
I am only reverting to `com.github.kristofa.brave` (I guess that's v3?) since it is used to propagate spans everywhere else, and there is a way to flush annotations without v4. I wanted to make it more consistent and reduce the number of dependencies in the interceptors, but that's debatable since it looks like a step backward for keeping things update-to-date.

**Comment by llinder** [comment]
@edwardsoo-ss thanks for describing the change.

Historically Zipkin and Brave have both moved at a very fast pace.  Right now there is a big push on the new Brave APIs.  For example there is work to introduce in process context propagation here https://github.com/openzipkin/brave/pull/369/files#diff-a68d2b3a2fe22417b7d89e982ab27cba  

We can revert to the more stable older version if others feel strongly about it.  That said I do fully expect to see more internal and external Brave code that starts using these new APIs so I'm comfortably with using them now.

The good thing about the Brave v4 APIs is that there is an adapter for backwards compatibility.  As you pointed out that does mean another dependency but I don't see a way around that during a transitional period anyway.

**Comment by edwardsoo-ss** [comment]
@llinder ok, I will update the change keeping v4

**Comment by edwardsoo-ss** [comment]
@llinder updated, another look please?

**Comment by llinder** [comment]
Sorry for not keeping on top of this.  @edwardsoo-ss do you think this is something we still want to merge?

I have a fairly big change coming and it will effect this a bit.  Mainly we just need to remove all Brave 3.x code which there is still some here.

If we want to get this in first I will take a look through this again tomorrow and will merge before my changes go in https://github.com/SmartThingsOSS/smartthings-brave/pull/8

File: brave-kafka/src/main/java/com/github/kristofa/brave/SpanUtils.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * the License.
+ */
+package com.github.kristofa.brave;
+
+import brave.propagation.TraceContext;
+
+/**
+ * Contains some util functions to get around some access modifiers
+ */
+public class SpanUtils {
```
Comment (llinder) [inline brave-kafka/src/main/java/com/github/kristofa/brave/SpanUtils.java]: SpanId is part of the now deprecated Brave 3.x APIs.  Just pointing this out as something we need to cleanup.  That can happen in another PR such as https://github.com/SmartThingsOSS/smartthings-brave/pull/8

File: brave-kafka/src/main/java/smartthings/brave/kafka/consumers/BaseTracingConsumerInterceptor.java [hunk 2]
```diff
@@ -89,6 +85,19 @@
...
   }
 
+  protected void reportServerReceivedSpan(TraceContext traceContext, ConsumerRecord<K, byte[]> record) {
+    // Only report "sr" if the span was sampled on client side;
+    // reporting only "sr" will result in NaN duration in the zipkin UI
+    if (traceContext != null && traceContext.sampled() == Boolean.TRUE) {
+      tracer.joinSpan(traceContext)
+        .name(nameProvider.spanName(record))
+        .remoteEndpoint(remoteEndpoint)
+        .annotate(Constants.SERVER_RECV)
```
Comment (llinder) [inline brave-kafka/src/main/java/smartthings/brave/kafka/consumers/BaseTracingConsumerInterceptor.java]: Setting this annotation explicitly shouldn't be necessary if you set the kind `Span.Kind.SERVER` and call call `start()`.  Also I wonder if the NaN duration is related to not calling `start()`?  Or maybe thats just a bug in the UI.

**Comment by llinder** [comment]
Just a note on the tests.  I think it would be really beneficial to create an integration test at some point.  There are several decent patterns for this such as using TestContainers http://testcontainers.viewdocs.io/testcontainers-java/ which work equally as well for CI as local development.

**Comment by edwardsoo-ss** [comment]
@llinder thanks for reviewing but let's close this one for now so it won't affect your big change; our traced kafka consumer service has sample rate set to 0 right now so this change is not too important right now

**Pull request closed without merging**